astropy/wcs/utils.py

```diff
@@ -825,3 +826,237 @@ def pixel_to_pixel(wcs_in, wcs_out, *inputs):
+      objects passed in for `world_coords` and `proj_point`, the frame for
+      `world_coords`  will override as the frame for the output WCS.
+
+    Parameters
+    ----------
+    xy : tuple of two `numpy.ndarray`
+        x & y pixel coordinates.
+    world_coords : `~astropy.coordinates.SkyCoord`
+        Skycoord object with world coordinates.
+    proj_point : 'center' or ~astropy.coordinates.SkyCoord`
+        Defaults to 'center', in which the geometric center of input world
+        coordinates will be used as the projection point. To specify an exact
+        point for the projection, a Skycoord object with a coordinate pair can
+        be passed in. For consistency, the units and frame of these coordinates
+        will be transformed to match 'world_coords' if they don't.
+    projection : str or `~astropy.wcs.WCS`
+        Three letter projection code, of any of standard projections defined 
+        in the FITS WCS standard. Optionally, a WCS object with projection 
+        keywords set may be passed in. 
+    sip_degree : int
+        If set to a non-zero integer value, will fit SIP of degree `sip_degree` 
+        to model geometric distortion. 
+
+    Returns
+    -------
+    wcs : `~astropy.wcs.WCS`
+        The best-fit WCS to the points given.
+    """
+
+    from astropy.coordinates import SkyCoord # here to avoid circular import
+    import astropy.units as u
+    from scipy.optimize import least_squares
+    from .wcs import Sip
+    import copy
+
+    xp, yp = xy
+    lon, lat = world_coords.data.lon.deg, world_coords.data.lat.deg
+
+    # verify input
+    if (proj_point != 'center') and (type(proj_point) != type(world_coords)):
+        raise ValueError("proj_point must be set to 'center', or an" +
+                         "`~astropy.coordinates.SkyCoord` object with " +
+                         "a pair of points.")
+    if proj_point != 'center':
+        assert proj_point.size == 1
+
+    proj_codes = [
+        'AZP', 'SZP', 'TAN', 'STG', 'SIN', 'ARC', 'ZEA', 'AIR', 'CYP',
+        'CEA', 'CAR', 'MER', 'SFL', 'PAR', 'MOL', 'AIT', 'COP', 'COE',
+        'COD', 'COO', 'BON', 'PCO', 'TSC', 'CSC', 'QSC', 'HPX', 'XPH'
+    ]
+    if type(projection) == str:
+        if projection not in proj_codes:
+            raise ValueError("Must specify valid projection code from list of " 
+                             + "supported types: ", ', '.join(proj_codes))
+        # empty wcs to fill in with fit values
+        wcs = celestial_frame_to_wcs(frame=world_coords.frame,
+                                     projection=projection)
+    else: #if projection is not string, should be wcs object. use as template.
+        wcs = copy.deepcopy(projection)
+        wcs.cdelt = (1., 1.) # make sure cdelt is 1
+        wcs.sip = None
+
+    # Change PC to CD, since cdelt will be set to 1
+    if wcs.wcs.has_pc():
+        wcs.wcs.cd = wcs.wcs.pc
+        wcs.wcs.__delattr__('pc')
```

Comment:
Internally wcslib works with the PC, regardless of whether the input has CD. I wouldn't try to delete it (it probably won't anyway). When you assign to wcs.cd the flag is set automatically so that has_cd() returns True and has_pc() returns False. The above three lines are not necessary.

Reply:
There was some discussion previously about changing the output to only have CD, since it was returning PC before. Should I change it back?

Reply:
I think right now it's not important how consistent the output is on this.  So as long as the resulting wcs is right, we can deal with whether there should be a cd/pc cleanup step later (which I *think* means you can delete this based on what @nden said, but hopefully the tests will let you check it)

Reply:
so we don't care if the output has a PC, even though cdelt will always be 1? or we dont care if it has a cd AND a pc (which should be the same, so just redundant info) ?

Reply:
Assign to the CD as before but don't delete the PC. I don't think it deletes it anyway and may cause only harm. When you assign `wcs.wcs.cd=pc` the PC is assigned it to CD and `wcs.wcs.has_cd()` returns `True`. However internally the PC is also set to that value and the library always uses the PC. This is just a feature of the internal library. So don't delete the PC.

Reply:
Ok I changed that - I assigned the object a CD that is the same as it's PC. However, when you print out the contents of the object it displays the PC by default, so it's not clear that this even has a CD so it seems useless to do that. Is there necessarily an issue with an object not having a PC, if it will be created from the CD internally anyway? When you construct a WCS from fits data that only has CD keywords, has_pc returns False and everything works fine. I just want to make sure because there was a change request to have the output have CD instead of PC. It could also complicate things more to have both - if the user wanted to manually reset a value in the object that has PC and CD, and they do it to just CD, the changes won't be reflected.

Reply:
wcslib works internally with the PC. Try this to see what I mean:
```
w = wcs.WCS()
w.wcs.cd=[[2,0],[0,2]]
print(w.wcs)
```
This prints the values of the wcslib struct and the PC is set there. It's probably OK to delete it as you do because I don't think it deletes it from the struct but I'm not sure if it has undesirable side effects. The safest thing to do is to create a new `WCS()` object and populate it from the user supplied one. But of course this defeats the purpose of supplying one as an argument because then you have to find what the projection parameters are and how they are stored in the header.

If I am reading the code correctly these assignment does not matter. The linear fit (line 1067) does not use these values as initial values, it uses the identity matrix. So essentially the only thing that is used from the supplied wcs is the projection. Unless I'm missing something. This doesn't seem like a very robust approach in the general case. But it will probably work. Sorry I'm not giving a concrete suggestion. I would probably not do it this way and would probably put some more burden on the user to specify the projection type. Since everyone else feels like this is the way to go once the other commentas are addressed this should probably be merged.

Reply:
@nden - to be clear on this 

> So essentially the only thing that is used from the supplied wcs is the projection

I think the other thing it's doing is that it is *also* using some of the random other parameters like `LONPOLE` if they are specified right?  That was the use case I was thinking of for why you'd pass in your own WCS.

Reply:
@eteq I believe `LONPOLE` is determined internally by the type of projection.  `LATPOLE` OTH may be set, though that's rare. Anyway I don't want to hold this up, when tests pass it should be merged.

Reply:
If 'projection' is passed in as a WCS, it will also reflect any non polynomial distortion in there in the fit, which may be useful in some cases. I purposefully don't have it use CD/PC from the input WCS as an initial guess, because if the refitting is being done because something is way off then the fit might be poor. I am going to go ahead and change this though, so that it does use input CD/PC as an initial guess, and i will just put a warning in the docstring of this potential behavior.  

I agree, I think I would prefer in general to have the user specify the projection with keywords alone since there is the already the function `celestial_frame_to_wcs` that should set up a WCS with all the correct keywords based on the projection type, and NOT have an input wcs as an option for projection. However, this still works and I think we should just go ahead with this since its late in the game and I'll just fix the above issue and the pep8 stuff.

Reply:
That said I agree it would be good to be able to set the initial value!